Modified OSD stats functions to accomodate single page stats for HD

M0j0Risin opened the pull request:
![SnapShot](https://user-images.githubusercontent.com/100724300/209537324-ed52903f-7820-41c3-9b4c-d7ffe08f75b3.jpg)

Implements a single function "osdShowStats()" for showing stats with parameters for whether the display is HD and which page to show if it is SD. So, existing paging functionality will be maintained for SD displays and a single stats page will be shown on HD displays.

M0j0Risin committed:
Modified OSD stats functions to accomodate single page stats for HD displays.

M0j0Risin committed:
Cleaned up curly brace styling to match coding standards

M0j0Risin committed:
Modified osdShowStats() to call the parameter "isSinglePageCompatible" and added a function osdDisplayIsSinglePageCompatible() to check if the screen size is at least the size for DJIWTF.

M0j0Risin committed:
Additional updates to support HDZERO.

Updated the function which decides whether to support single page stats, osdVideoSystemIsSinglePageStatsCompatible(), to be based on the enumerated value for the selected video system.

Combined MAX SPEED and AVG SPEED on to one line.

Combined MIN RSSI % and MIN RSSI DBM on to one line.

So we're down to 16 lines, 17 including the blank line at the top.

M0j0Risin committed:
Modified 'MAX/AVG SPEED' to read 'SPEED MAX/AVG' to be more consistent with the other two multi-valued lines.

M0j0Risin committed:
Changed back to 'MAX/AVG SPEED', I don't know what I was seeing before. This is more consistent.

M0j0Risin committed:
Added temporary placeholder comment for the change necessary to support Avatar once it is added to the videoSystem_e enumeration.

M0j0Risin committed:
Fixed merge conflicts with PR #8439.

I also corrected some buggy behavior that was happening surrounding the stick commands to change pages for multi-page stats.

M0j0Risin committed:
Reintroduced logic to refresh single-page stats on multi-page change events so that the "Saving Settings" and "Settings Saved" messages have a chance to be displayed since it wouldn't redraw them normally.

M0j0Risin committed:
Cleaned up an unnecessary comment block I inadevertently left in there.

MrD-RC commented on src/main/io/osd.c:
```diff
@@ -4239,57 +4364,60 @@ static void osdRefresh(timeUs_t currentTimeUs)
             osdSetNextRefreshIn(delay);
         } else {
-            osdShowStatsPage1(); // show first page of statistics
+            osdShowStats(osdVideoSystemIsSinglePageStatsCompatible(), 1); // show first page of statistics
```
Just a thought. Instead of having that function, you could just use this
`osdShowStats((osdDisplayPort->rows >= 18), 1);`
Though this won't explicitly check for HD video systems, it will make sure that the height of the OSD is adequate. Even with the minimum ratio of 4:3, that would be 32 characters wide (The current single page stats would fit). But currently there are no 4:3 HD displays at 18 characters high. So 50 characters max width would be fine. Even HDZero in 4:3 mode uses the full 16:9 for the stats screen.

M0j0Risin replied:
I like that that. It makes sense and would eliminate the need to modify this again to accommodate Avatar.  I'll go ahead and do that.  I may add a constant to hold that 18 value. Something like MAX_LINES_FOR_SINGLE_PAGE or whatever.

M0j0Risin replied:
I made the changes you suggested. Detection of compatablility for single-page stats is now based strictly on the number of rows currently displayed. OSD_STATS_SINGLE_PAGE_MIN_ROWS stores the minimum number of rows for single-page stats.

I also made some additional improvements surrounding the logic for manually paging using the stick commands for multi-page stats. I think this makes the code a lot easier to follow. 

There is one caveat surrounding the timing of when "Saved Settings" is displayed. Since it is only displayed when osdShowStats() is called, it could be missed if the 'stats_page_auto_swap_time' is 0, greater than 4 seconds or if the user has used the stick commands to manually page with multi-page stats since it effectively pauses the updates. So multi-page users would only see "Saved Settings" in that case if they happen to manually go to the next page within the 5 second window. Probably no big deal, but I figured I'd mention that. I also added this in the comments in the code for clarity.

MrD-RC commented on src/main/io/osd.c:
```diff
@@ -3867,242 +3911,276 @@ static void osdUpdateStats(void)
+        displayWrite(osdDisplayPort, statNameX, top, "DISARMED BY      :");
+        displayWrite(osdDisplayPort, statValuesX, top++, disarmReasonStr[getDisarmReason()]);
+
+        /* We only want to show this message once, so it will be shown after the second
```
Could this *if* not be used after the *page 2* part has been drawn? That would save the code duplication, and it would be drawn once on single page and on each page of two page stats.

Just to clarify, I'm meaning this if statement.
```
if (savingSettings == true) {
    displayWrite(osdDisplayPort, statNameX, top++, OSD_MESSAGE_STR(OSD_MSG_SAVING_SETTNGS));
} else if (notify_settings_saved > 0) {
    if (millis() > notify_settings_saved) {
        notify_settings_saved = 0;
    } else {
        displayWrite(osdDisplayPort, statNameX, top++, OSD_MESSAGE_STR(OSD_MSG_SETTINGS_SAVED));
    }
}
```

M0j0Risin replied:
Yeah, I think you're right there. It should be after both if blocks, duh.  I should have realized that.  I'll make that change as well.

M0j0Risin replied:
See my comment above, but this change is also included in my latest commit.

MrD-RC reviewed (commented):
I have a couple of suggestions. See what you think?

M0j0Risin committed:
Merge branch 'iNavFlight:master' into Single-page-stats-for-HD-displays

M0j0Risin committed:
Detection of compatablility for single-page stats is now based strictly on the number of rows currently displayed. OSD_STATS_SINGLE_PAGE_MIN_ROWS stores the minimum number of rows for single-page stats.

I also made some additional improvements surrounding the logic for manually paging using the stick commands for multi-page stats. I think this makes the code a lot easier to follow.
